nucliadb_utils/nucliadb_utils/nats/demux.py

```diff
@@ -0,0 +1,310 @@
+                return
+
+            await worker.add_work(work_unit)
+            self.working_on.add(work_id)
+
+            if created or (not created and not worker.working):
+                # The second condition should never happen, the worker task finished
+                # between us getting the worker and adding work. We'll create a new
+                # task for it
+                self.start_worker(split)
+
+            loop = asyncio.get_event_loop()
+            wait_condition = loop.create_future()
+
+            work_unit_done.add_listener(
+                work_id,
+                partial(
+                    NatsDemultiplexer.on_work_unit_done,
+                    work_id=work_id,
+                    notifier=wait_condition,
+                ),
+            )
+
+            work_outcome = await wait_condition
+            self.working_on.discard(work_id)
+            if work_outcome.error:
+                await msg.nak()
+            else:
+                if work_outcome.success:
+                    await msg.ack()
+                else:
+                    await msg.nak()
+
+    def get_or_create_worker(self, worker_id: str) -> tuple["DemuxWorker", bool]:
+        new_worker = worker_id not in self.workers
+        if new_worker:
+            queue = self.queue_klass()
+            worker = DemuxWorker(queue, self.processor.process)
+            self.workers[worker_id] = worker
+        else:
+            worker = self.workers[worker_id]
+        return (worker, new_worker)
+
+    def start_worker(self, worker_id: str):
+        worker = self.workers[worker_id]
+        task = asyncio.create_task(worker.work_until_finish())
+        task.add_done_callback(partial(self.demux_worker_done, worker_id))
+        return task
+
+    def demux_worker_done(self, worker_id: str, task: asyncio.Task):
+        self.workers.pop(worker_id, None)
+
+        # propagate errors
+        if task.exception() is not None:
+            raise task.exception()  # type: ignore
+
+    @staticmethod
+    async def on_work_unit_done(
+        work_id: str,
+        notifier: asyncio.Future,
+        payload: WorkOutcome,
+    ):
+        if payload.work_id == work_id:
+            notifier.set_result(payload)
+
+
+class DemuxWorker:
+    def __init__(
+        self,
+        work_queue: asyncio.Queue,
+        cb: Callable[[DemuxProcessorWork], Awaitable[bool]],
+    ):
+        self.cb = cb
+        self.work_queue = work_queue
+        self.working = False
+
+    async def add_work(self, work: WorkUnit):
+        await self.work_queue.put(work)
+
+    async def work_until_finish(self):
+        self.working = True
+        while not self.work_queue.empty():
+            work: WorkUnit = await self.work_queue.get()
+            await self.do_work(work)
+        self.working = False
+
+    async def do_work(self, work: WorkUnit):
+        try:
+            logger.info(f"Working on '{work.work_id}'", extra={"work_id": work.work_id})
```

Comment:
for logging purposes, maybe we want to have the slipt logged too? or the callback func name.